src/io/ioexternalbus.cpp

```diff
@@ -28,10 +50,217 @@ QString IOExternalBus::description() const {
+            dprintf("write ok\n");
+        } else {
+            skt_use = 0;
+            printf("write error !!!!!!!!!!!!!!!!!!!!!!\n");
+            tcpSocket->close();
+        }
+        skt_use = 0;
+    }
+}
+
+void IOExternalBus::connectButtonTriggered() {
+    if (m_ui->connectButton->text().contains("Connect")) {
+        tcpSocket->abort();
+        if (tcpSocket->connectToHost(m_ui->address->text(), m_ui->port->value())) {
+            if (send_cmd(VB_PINFO) < 0) {
+                disconnectOnError();
+                return;
+            }
+            cmd_header_t cmd_header;
+
+            if (recv_cmd(cmd_header) < 0) {
+                disconnectOnError();
+                return;
+            }
+            if (cmd_header.payload_size) {
+                QByteArray buff(cmd_header.payload_size + 1, 0);
+                if (recv_payload(buff, cmd_header.payload_size) < 0) {
+                    disconnectOnError();
+                    return;
+                }
+                dprintf("%s\n", buff.data());
+
+                QJsonParseError error;
+                QJsonDocument desc = QJsonDocument::fromJson(buff.data(), &error);
+                if (error.error == QJsonParseError::NoError) {
+                    const unsigned int addrw = desc.object().value(QString("address width")).toInt();
+                    QJsonObject osymbols = desc.object().value(QString("symbols")).toObject();
+
+                    m_regDescs.clear();
+                    for (QJsonObject::iterator i = osymbols.begin(); i != osymbols.end(); i++) {
+                        m_regDescs.push_back(
+                            RegDesc{i.key(), RegDesc::RW::RW, addrw * 8, static_cast<AInt>(i.value().toInt()), true});
+                    }
+
+                    ByteSize = addrw * osymbols.count();
+
+                    m_ui->connectButton->setText("Disconnect");
+                    m_ui->status->setText("Connected");
+                    m_ui->server->setText(desc.object().value(QString("name")).toString());
+
+                    emit regMapChanged();
+                    emit sizeChanged();
+
+                    // delete[] buff;
+                } else {
+                    QMessageBox::information(nullptr, tr("Ripes VBus"), QString("json: ") + error.errorString());
+                    m_ui->connectButton->setText("Connect");
+                    m_ui->status->setText("Disconnected");
+                    m_ui->server->setText("-");
+
+                    send_cmd(VB_QUIT);
+                    tcpSocket->abort();
+                }
+            }
+        } else {
+            disconnectOnError();
+        }
+    } else {  // disconnect
+        m_ui->connectButton->setText("Connect");
+        m_ui->status->setText("Disconnected");
+        m_ui->server->setText("-");
+
+        send_cmd(VB_QUIT);
+        tcpSocket->abort();
+
+        m_regDescs.clear();
+
+        emit regMapChanged();
+        emit sizeChanged();
```

Comment:
As above, could be moved to the end of the function and always emitted.

Reply:
Moved.